File: solutions/fsharp/sum-of-multiples/1/SumOfMultiples.fs
module SumOfMultiples

let sumOfMultiples (numbers: int list) (upperBound: int): int =
  if upperBound > 0 then 
    let range = [1..(upperBound-1)]

    numbers
    |> List.map(
      fun x -> range |> List.filter(fun y -> y % x = 0))
    |> List.collect(fun x -> x)
    |> List.distinct 
    |> List.sum
  else
    0

File: solutions/haskell/sum-of-multiples/1/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.List(nub)

-- determine all the multiples of a number that are less than the limit
sumElems :: Integer -> Integer -> [Integer]
sumElems n limit 
    | n > 0 = takeWhile (< limit) (map (* n) [1..])
    | otherwise = []

-- collect all the multiples, merge together in a sinle list and remove duplicates
collectElems :: [Integer] -> Integer -> [Integer] -> [Integer]
collectElems [] _ acc = nub acc  -- get only unique elements
collectElems (x:xs) limit acc = collectElems xs limit newacc
                                where newacc = sumElems x limit ++ acc

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples xs limit = sum $ collectElems xs limit []